libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp

```diff
@@ -35,89 +57,138 @@ AP_Volz_Protocol::AP_Volz_Protocol(void)
-        update_volz_bitmask(bitmask);
+    // Create thread to handle output
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Volz_Protocol::loop, void),
+                                          "Volz",
+                                           1024, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
+        AP_BoardConfig::allocation_error("Volz thread");
     }
+}
 
-    uint32_t now = AP_HAL::micros();
-    if (now - last_volz_update_time < volz_time_frame_micros ||
-        port->txspace() < VOLZ_DATA_FRAME_SIZE) {
-        return;
-    }
+void AP_Volz_Protocol::loop()
+{
+    port->begin(115200, UART_BUFSIZE_RX, UART_BUFSIZE_TX);
+    port->set_unbuffered_writes(true);
+    port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+
+    // Calculate the amount of time it should take to send a command
+    // B/s to s/B, 1000000 converts to microseconds, multiply by number of bytes
+    // 6 bytes at 11520 bytes per second takes 520 us
+    const uint16_t send_us =  (sizeof(CMD) * 1000000) / port->bw_in_bytes_per_second();
 
-    last_volz_update_time = now;
+    // receive packet is same length as sent, double to allow some time for the servo respond
+    const uint16_t receive_us = send_us * 2;
 
-    uint8_t i;
-    uint16_t value;
+    // This gives a total time of 1560ms, message rate of 641 Hz.
+    // One servo at 641Hz, two at 320.5 each, three at 213.7 each ect...
 
-    // loop for all channels
-    for (i=0; i<NUM_SERVO_CHANNELS; i++) {
-        // check if current channel is needed for Volz protocol
-        if (last_used_bitmask & (1U<<i)) {
+    while (port != nullptr) {
 
-            SRV_Channel *c = SRV_Channels::srv_channel(i);
-            if (c == nullptr) {
+        // Wait the expected amount of time for the send and receive to complete so we don't step on the response
+        hal.scheduler->delay_microseconds(send_us + receive_us);
+
+        while (port->txspace() < sizeof(CMD)) {
+            // Wait until there is enough space to transmit a full command
+            hal.scheduler->delay_microseconds(10);
+        }
+
+        // loop for all channels
+        for (uint8_t i=0; i<ARRAY_SIZE(servo.pwm); i++) {
+            // Send each channels in turn
+            const uint8_t index = (last_sent_index + 1 + i) % ARRAY_SIZE(servo.pwm);
+            if ((uint32_t(bitmask.get()) & (1U<<index)) == 0) {
+                // Not configured to send
                 continue;
             }
-            
-            // check if current channel PWM is within range
-            if (c->get_output_pwm() < VOLZ_PWM_POSITION_MIN) {
-                value = 0;
-            } else {
-                value = c->get_output_pwm() - VOLZ_PWM_POSITION_MIN;
+            last_sent_index = index;
+
+            uint16_t pwm;
+            {
+                // Get PWM from saved array
+                WITH_SEMAPHORE(servo.sem);
+                pwm = servo.pwm[index];
```

Comment:
16 bit number is atomic anyway, so semaphore doesn't help

Reply:
Done.